woocommerce/woocommerce-ios
# [Product Multi-Selection] Add `RemoteOrderSynchronizer` support for multiple products

Closes: #8923 

## Description
As part of Product Multi-Selection M1.1, we'll be sending an unique request to the remote with all products that conform an Order, for this we need to update the `RemoteOrderSynchronizer` to accept multiple products.

## Changes
- Updates the `RemoteOrderSynchronizer` and `OrderSynchronizerprotocol` with a new `PassthroughSubject` variable, which accepts an `[OrderSyncProductInput]` array as Output.
- Inside this subject, called `setProducts()`, we'll update the Order that has to be sync by transforming inputs as necessary, following the same implementation as the existing `setProduct()`. We use those to update the underlying Order as inputs are received.
- Add Unit Tests

Question: The static methods `update()` and `updateOrderItems()` are pretty much identical, with the difference of the former returning an `Order` object, and the latter retuning an `[OrderItem]` object. I found this distinction necessary when updating the multiple `OrderSyncProductInputs` into a single `Order` and couldn't simplify it into a single method without extensive changes. Happy to get any feedback on this so I can tackle it here or on a different PR

## Testing instructions
- Unit Tests should pass
- We can also test this UI wise, **please note that this is not the final behaviour**, but just to see if works as expected:
1. In [EditableOrderViewModel](https://github.com/woocommerce/woocommerce-ios/blob/trunk/WooCommerce/Classes/ViewRelated/Orders/Order%20Creation/EditableOrderViewModel.swift#L731), change the following line inside the `addProductToOrder()` method :
```
//orderSynchronizer.setProduct.send(input)
orderSynchronizer.setProducts.send([input, input, input, input, input])
```
2. Run the app. Go to Orders > `+` > `Add Products` > Tap on any product > See how 5 products of the same type are added to the Order.

![Simulator Screen Recording - iPhone 13 Pro - 2023-03-16 at 19 45 41](https://user-images.githubusercontent.com/3812076/225621376-4a8681ee-45e0-418b-9a7b-0902973f99ef.gif)
